stor/dx.py

```diff
@@ -0,0 +1,513 @@
+                  category=None):
+        """Iterates over listed files that match an optional pattern.
+
+        Args:
+            pattern (str): glob pattern to match the filenames against.
+            canonicalize (boolean): whether to return canonicalized paths
+            recurse (boolean): whether to look in subfolders of folder as well
+            starts_with (str): Allows for an additional search path to
+                be appended to the resource of the dx path. Note that this
+                resource path is treated as a directory
+            limit (int): Limit the amount of results returned
+            category (str): Restricting class : One of 'record', 'file', 'gtable,
+                'applet', 'workflow'
+
+        Returns:
+             Iter[DXPath]: Iterates over listed files that match an optional pattern.
+        """
+        proj_id = self.canonical_project
+        proj_name = self.virtual_project
+        kwargs = {
+            'project': proj_id,
+            'name': pattern,
+            'name_mode': 'glob',
+            # the query performance is similar w/wo describe field,
+            # hence no need to customize query based on canonicalize flag
+            'describe': {'fields': {'name': True, 'folder': True}},
+            'recurse' : recurse,
+            'classname': category,
+            'limit': limit,
+            'folder': (self.resource or '/') + (starts_with or '')
+        }
+        if self.resource and not self._is_folder():  # if path is a file path
+            yield []
+            return
+        list_gen = dxpy.find_data_objects(**kwargs)
+        for obj in list_gen:
+            if canonicalize:
+                yield DXCanonicalPath('dx://{}:/{}'.format(obj['project'], obj['id']))
+            else:
+                dx_p = DXVirtualPath(self.drive + proj_name + ':' + obj['describe']['folder'])
+                dx_p = dx_p / obj['describe']['name']
+                yield dx_p
+
+    def glob(self, pattern, condition=None, canonicalize=False):
+        """ Glob for pattern relative to this directory."""
+
+        results = list(self.walkfiles(
+            canonicalize=canonicalize,
+            pattern=pattern
+        ))
+        if not results or not results[0]:  # when results == [[]]
+            results = []
+        utils.validate_condition(condition)
+        utils.check_condition(condition, results)
+        return results
+
+    def exists(self):
+        """Checks existence of the path.
+
+        Returns True if the path exists, False otherwise.
+
+        Returns:
+            bool: True if the path exists, False otherwise.
+        """
+        try:
+            # first see if there is a specific corresponding object
+            self.stat()
+            return True
+        except (NotFoundError, ValueError):
+            pass
+        # otherwise we could be a directory, so try to grab first
+        # file/subfolder
+        if self._is_folder():
```

Comment:
can you help me understand what the `_is_folder` check adds here? why not just try to list it directly and return False on the NotFoundError?

Reply:
can you ensure there's an integration test that shows the following is True?

```
# empty dirs should be seen as existing
folder_path = dx_p / 'myfolder'
folder_path.makedirs_p()
assert stor.exists(folder_path)
```

Reply:
should be a documented diff between swift vs. s3 vs. posix

Swift & S3 - "empty" folder is exists = False
Posix & DX - "empty" folder is exists = True

Reply:
Removed _is_folder

There is a test to check that empty dirs are seen as existing. Should I add an integration test as well?